capstone/src/main/java/com/google/sps/Objects/QuadTree.java

```diff
@@ -0,0 +1,211 @@
+    double yMin = bounds.getYMin();
+    double xMax = bounds.getXMax();
+    double yMax = bounds.getYMax();
+    double xOffset = xMin + (xMax - xMin) / 2;
+    double yOffset = yMin + (yMax - yMin) / 2;
+
+    /*
+    * Split space into 4 quadrant trees and create new bounding boxes 
+    * 
+    *xMax    ************************
+    *        *          *           *
+    *        * topLeft  *  topRight *
+    *        *          *           *
+    *xOffset ************************ 
+    *        *          *           *
+    *        * botLeft  *  botRight *
+    *        *          *           *
+    *xMin    ************************
+    *        yMin     yOffset     yMax
+    */
+    botRightTree = new QuadTree(NODE_CAPACITY, level + 1, 
+        new BoundingBox(xMin, yOffset, xOffset, yMax));
+    topRightTree = new QuadTree(NODE_CAPACITY, level + 1, 
+        new BoundingBox(xOffset, yOffset, xMax, yMax));
+    botLeftTree = new QuadTree(NODE_CAPACITY, level + 1, 
+        new BoundingBox(xMin, yMin, xOffset, yOffset));
+    topLeftTree = new QuadTree(NODE_CAPACITY, level + 1, 
+        new BoundingBox(xOffset, yMin, xMax, yOffset));
+
+    children.add(topRightTree);
+    children.add(topLeftTree);
+    children.add(botLeftTree);
+    children.add(botRightTree);
+  }
+
+  public Location nearestNeighbor(Location loc) {
+    Location firstInList = this.locations.get(0);
+    return nearestNeighbor(loc, firstInList);
+  }
+
+  public Location nearestNeighbor(Location loc, Location closest) {
+
+    double closestDistance = euclidianDistance(loc, closest);
+
+    // Only check quadrants within radius of loc and closest
+    if (!bounds.intersectsCircle(loc, closestDistance)) {
+      return closest;
+    }
+
+    // Check distance between location and points in this quadrant
+    for (Location location: locations) {
+      if (euclidianDistance(loc, location) < closestDistance) {
+        closest = location;
+        closestDistance = euclidianDistance(loc, location);
+      }
+    }
+
+    // Check distance between location and points within all children
+    if (!(topLeftTree == null)) {
+      for (QuadTree qt : children) {
+        closest = qt.nearestNeighbor(loc, closest);
+      }
+    }
+    return closest;
+  }
+
+  public ArrayList<Location> findKNearestNeighbors(Location loc, int k) {
+    /*
+    * Create priority queue of size k ordered by distance between point and 
+    * midpoint 
+    * Last point is the maxClosest distance 
+    */
+    PriorityQueue<Location> pq = new PriorityQueue<Location>(k, Comparator.comparingDouble( locationPQ -> -euclidianDistance(loc, locationPQ)));
+
+    pq = findKNearestNeighborsHelper(loc, k, pq);
+
+    ArrayList<Location> nearestLocations = new ArrayList<>();
+
+    while (!pq.isEmpty()) {
+      nearestLocations.add(0, pq.poll());
+    }
+
+    return nearestLocations;
+  }
+
+  public PriorityQueue<Location> findKNearestNeighborsHelper(Location loc, int k, PriorityQueue<Location> pq) {
```

Comment:
instead of `pq` I'd consider `closestLocationPQ` or `nearestPQ` or `farthestPQ` something that tells me what is being prioritized